tap_sftp/client.py

```diff
@@ -72,8 +73,11 @@ def sftp(self, sftp):
         self.__sftp = sftp
 
     def close(self):
-        self.sftp.close()
-        self.transport.close()
+        if hasattr(self, 'sftp') and self.sftp is not None:
+            self.sftp.close()
+
+        if self.transport is not None:
+            self.transport.close()
```

Comment:
What do you think of this: https://github.com/Shopify/tap-sftp-xml-support/pull/6/files

Reply:
I think the part you're missing is that the SFTP connection doesn't exits yet in the error I'm try/catching:

```
try:
            self.transport = paramiko.Transport((self.host, self.port))
            self.transport.use_compression(True)
            self.transport.connect(username=self.username, password=self.password, hostkey=None, pkey=self.key)
            self.__sftp = paramiko.SFTPClient.from_transport(self.transport)
        except (AuthenticationException, SSHException) as ex:
            LOGGER.warning('Connection attempt failed: %s', ex)
            self.close()
            raise
```

when this fail, the self.__sftp variable is not set and the client isn't created. Thus I can't call the client `close` method that you're referring to.

Reply:
additional context: if we don't use the `hasattr(self, 'sftp')` bit & we get into the exception block, I see this error:

```AttributeError: 'SFTPConnection' object has no attribute '_SFTPConnection__sftp'`

This tells me that at this point, the client hasn't been successfully created and we haven't set the `self.__sftp` variable.

If we try to close the client, it gets ignored because (luckily) we have the `hasattr` to check if the attribute exists. If our intention is to close the `Transport` by relying on closing the client, it won't work.

Reply:
Yeah. The SFTPConnection class is a bit of a head scratcher. I don't know why sftp is an @property since it appears to never be used and seems better encapsulated as a private member. I know this is a fork of some other code. :thinking_face

But, if feels like once we remove the unnecessary sftp property, it simplifies to:

```
    def __init__(self, host, username, password=None, private_key_file=None, port=None):
        ...
        self.sftp = None
        ...

    def _attempt_connection(self):
        try:
            transport = paramiko.Transport((self.host, self.port))
            transport.use_compression(True)
            transport.connect(username=self.username, password=self.password, hostkey=None, pkey=self.key)
            self.sftp = paramiko.SFTPClient.from_transport(self.transport)
        except (AuthenticationException, SSHException) as ex:
            LOGGER.warning('Connection attempt failed: %s', ex)
            self.close()
            raise

    def close(self):
        if self.sftp:
            self.sftp.close()
        if self.decrypted_file:
            self.decrypted_file.close()
```